## Preprocessing

In [5]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("./Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
pip install tensorflow

     |████████████████████████████████| 244.3 MB 3.5 kB/s eta 0:00:015    |████████▏                       | 62.3 MB 5.4 MB/s eta 0:00:34     |████████████▉                   | 98.0 MB 3.9 MB/s eta 0:00:38.3 MB 3.3 MB/s eta 0:00:39     |████████████████████████▏       | 184.8 MB 1.2 MB/s eta 0:00:49
     |████████████████████████████████| 1.7 MB 5.2 MB/s eta 0:00:01
     |████████████████████████████████| 57 kB 1.8 MB/s eta 0:00:01
     |████████████████████████████████| 1.6 MB 9.9 MB/s eta 0:00:01
     |████████████████████████████████| 126 kB 14.3 MB/s eta 0:00:01
     |████████████████████████████████| 65 kB 9.5 MB/s  eta 0:00:01
     |████████████████████████████████| 6.0 MB 4.1 MB/s eta 0:00:01     |████████████████████▎           | 3.8 MB 7.4 MB/s eta 0:00:01
     |████████████████████████████████| 439 kB 2.8 MB/s eta 0:00:01
     |████████████████████████████████| 25.0 MB 7.9 MB/s eta 0:00:01     |████████████████████████████████| 25.0 MB 7.9 MB/s 
     |████████████████████████

In [6]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'])
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [7]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [8]:
# Look at APPLICATION_TYPE value counts for binning
appSeries = application_df['APPLICATION_TYPE'].value_counts()
appSeries

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [9]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
app_cutoff = 500

application_types_to_replace = []

for item in appSeries.index:
    if appSeries[item]<app_cutoff:
        application_types_to_replace.append(item)
        
application_types_to_replace

['T9', 'T13', 'T12', 'T2', 'T25', 'T14', 'T29', 'T15', 'T17']

In [10]:


# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [11]:
# Look at CLASSIFICATION value counts for binning
classSeries = application_df['CLASSIFICATION'].value_counts()
classSeries

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [12]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#find list of values to drop
for item in classSeries.index:
    if classSeries[item] == 1:
        application_df = application_df[application_df.CLASSIFICATION != item]
        classSeries = classSeries.drop(item)
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C4100        6
C1720        6
C2400        6
C1600        5
C1257        5
C1260        3
C2710        3
C0           3
C1267        2
C1256        2
C1234        2
C1246        2
C3200        2
Name: CLASSIFICATION, dtype: int64

In [13]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
class_cutoff = 100

classifications_to_replace= []

for item in classSeries.index:
    if classSeries[item]<class_cutoff:
        classifications_to_replace.append(item)
        
classifications_to_replace 

['C2800',
 'C7100',
 'C1300',
 'C1280',
 'C1230',
 'C1400',
 'C7200',
 'C2300',
 'C1240',
 'C8000',
 'C7120',
 'C1500',
 'C1800',
 'C6000',
 'C1250',
 'C8200',
 'C1238',
 'C1278',
 'C1235',
 'C1237',
 'C7210',
 'C2400',
 'C1720',
 'C4100',
 'C1257',
 'C1600',
 'C1260',
 'C2710',
 'C0',
 'C3200',
 'C1234',
 'C1246',
 'C1267',
 'C1256']

In [14]:


# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      643
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64

In [15]:
# Convert categorical data to numeric with `pd.get_dummies`
type_list = application_df.dtypes
type_list

APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
STATUS                     int64
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object

In [16]:
application_df = pd.get_dummies(application_df, columns=type_list.index)
application_df

APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  \
0                           0                     1                     0   
1                           0                     0                     0   
2                           0                     0                     0   
3                           0                     0                     0   
4                           0                     0                     0   
...                       ...                   ...                   ...   
34294                       0                     0                     0   
34295                       0                     0                     0   
34296                       0                     0                     0   
34297                       0                     0                     0   
34298                       0                     0                     0   

       APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  \
0                        0                    0                    0   
1                        1                    0                    0   
2                        0                    0                    1   
3                        1                    0                    0   
4                        1                    0                    0   
...                    ...                  ...                  ...   
34294                    0                    1                    0   
34295                    0                    1                    0   
34296                    1                    0                    0   
34297                    0                    0                    1   
34298                    1                    0                    0   

       APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  APPLICATION_TYPE_T8  \
0                        0                    0                    0   
1                        0                    0                    0   
2                        0                    0                    0   
3                        0                    0                    0   
4                        0                    0                    0   
...                    ...                  ...                  ...   
34294                    0                    0                    0   
34295                    0                    0                    0   
34296                    0                    0                    0   
34297                    0                    0                    0   
34298                    0                    0                    0   

       AFFILIATION_CompanySponsored  ...  ASK_AMT_1893400128  \
0                                 0  ...                   0   
1                                 0  ...                   0   
2                                 1  ...                   0   
3                                 1  ...                   0   
4                                 0  ...                   0   
...                             ...  ...                 ...   
34294                             0  ...                   0   
34295                             1  ...                   0   
34296                             1  ...                   0   
34297                             0  ...                   0   
34298                             0  ...                   0   

       ASK_AMT_2264109450  ASK_AMT_2310256039  ASK_AMT_3391919220  \
0                       0                   0                   0   
1                       0                   0                   0   
2                       0                   0                   0   
3                       0                   0                   0   
4                       0                   0                   0   
...                   ...                 ...                 ...   
34294                   0                   0                   0   
34295                   0      

In [19]:
# Split our preprocessed data into our features and target arrays
y = application_df.IS_SUCCESSFUL_1.values
X = application_df.drop(columns="IS_SUCCESSFUL_1").values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

In [20]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [21]:
input_dimensions=X.shape[1]
input_dimensions

8790

## Compile, Train and Evaluate the Model

In [22]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_model = tf.keras.models.Sequential()

nn_model.add(tf.keras.layers.Dense(units=8, activation="relu", input_dim=input_dimensions))

# First hidden layer
nn_model.add(tf.keras.layers.Dense(units=8, activation="relu"))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(units=8, activation="relu"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 8)                 70328     
                                                                 
 dense_1 (Dense)             (None, 8)                 72        
                                                                 
 dense_2 (Dense)             (None, 8)                 72        
                                                                 
 dense_3 (Dense)             (None, 1)                 9         
                                                                 
Total params: 70,481
Trainable params: 70,481
Non-trainable params: 0
_________________________________________________________________


2023-02-01 18:56:46.563339: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [23]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [24]:
# Train the model
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 4s 4ms/step - loss: 0.3154 - accuracy: 0.8291
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.1171 - accuracy: 0.9505
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.0367 - accuracy: 0.9872
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.0152 - accuracy: 0.9946
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.0061 - accuracy: 0.9979
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.0035 - accuracy: 0.9987
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.0024 - accuracy: 0.9989
Epoch 8/100
804/804 [==============================] - 3s 3ms/step - loss: 0.0015 - accuracy: 0.9995
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.0012 - accuracy: 0.9996
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 8.5444e-04 - ac

804/804 [==============================] - 2s 3ms/step - loss: 6.7176e-05 - accuracy: 1.0000
Epoch 80/100
804/804 [==============================] - 2s 3ms/step - loss: 4.3457e-05 - accuracy: 1.0000
Epoch 81/100
804/804 [==============================] - 2s 2ms/step - loss: 3.0886e-05 - accuracy: 1.0000
Epoch 82/100
804/804 [==============================] - 2s 3ms/step - loss: 2.7669e-05 - accuracy: 1.0000
Epoch 83/100
804/804 [==============================] - 2s 3ms/step - loss: 9.4395e-06 - accuracy: 1.0000
Epoch 84/100
804/804 [==============================] - 2s 3ms/step - loss: 7.2120e-06 - accuracy: 1.0000
Epoch 85/100
804/804 [==============================] - 2s 3ms/step - loss: 1.2861e-04 - accuracy: 1.0000
Epoch 86/100
804/804 [==============================] - 2s 3ms/step - loss: 7.9379e-05 - accuracy: 1.0000
Epoch 87/100
804/804 [==============================] - 2s 3ms/step - loss: 5.5492e-05 - accuracy: 1.0000
Epoch 88/100
804/804 [==============================] - 2s 

In [26]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.0448 - accuracy: 0.9965 - 672ms/epoch - 3ms/step
Loss: 0.044759564101696014, Accuracy: 0.9964990019798279


In [38]:
# Export our model to HDF5 file
nn_model.save('AlphabetSoupCharity.h5')

# The model has achieved 99.6% accuracy, ahead of the 75% target.